### PR TITLE
feat(trigger): add new apig types for trigger

### LINKED ISSUE #1891: [FunctionGraph] Trigger does not support APIG type
Hi team,

According to the current usage analysis of some customers, the trigger needs to support the APIG type of the shared and the dedicated kind to trigger the function through a specific request.

Best regards!

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
There are currently two ways to trigger functions through the API, one of which is through triggers.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1891

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new APIG and DEDICATEDGATEWAY type for trigger.
```

## PR Checklist

* [ ] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
NONE
```

### CODE REVIEW

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 1]
```diff
@@ -66,7 +72,7 @@ func ResourceFunctionGraphTrigger() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 				ValidateFunc: validation.StringInSlice([]string{
-					timingTrigger, obsTrigger, smnTrigger, disTrigger, kafkaTrigger,
+					timingTrigger, obsTrigger, smnTrigger, disTrigger, kafkaTrigger, apigTrigger, dedicatedApigTrigger,
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:75]: pls add new valid values APIG and DEDICATEDGATEWAY for `type` in the docs

File: docs/resources/fgs_trigger.md [hunk 2]
```diff
@@ -261,6 +307,29 @@ The `kafka` block supports:
...
+
+* `api_name` - (Required, String, ForceNew) Specifies the API name. Changing this will create a new trigger resource.
+
+* `instance_id` - (Optional, String, ForceNew) Specifies the ID of the APIG dedicated instance to which the API belongs.
+  Required if the `type` is `DEDICATEDGATEWAY`. Changing this will create a new trigger resource.
+
+* `security_authentication` - (Optional, String, ForceNew) Specifies the security authentication mode. The valid values
+  are **NONE**, **APP** and **IAM**, default to **NONE**. Changing this will create a new trigger resource.
+
+* `request_protocol` - (Optional, String, ForceNew) Specifies the request protocol of the API. The valid value are
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md:327]: the default value is `HTTPS`, pls add description

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 3]
```diff
@@ -338,6 +403,116 @@ func buildKafkaEventData(d *schema.ResourceData) map[string]interface{} {
...
+
+	return envId, subDomain, nil
+}
+
+// Obtain environment ID and sub-domain of shared APIG.
+func getDedicatedApigSubDomainAndEnvId(d *schema.ResourceData, config *config.Config) (string, string, error) {
+	var envId, subDomain string
+	apigwClient, err := config.ApigV2Client(config.GetRegion(d))
+	if err != nil {
+		return envId, subDomain, fmtp.Errorf("Error creating HuaweiCloud shared APIG v1.0 client: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: s/v1.0/v2

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 4]
```diff
@@ -492,6 +674,37 @@ func setKafkaEventData(d *schema.ResourceData, eventData map[string]interface{})
...
+			"env_name":                eventData["env_name"],
+			"security_authentication": eventData["auth"],
+			"request_protocol":        eventData["protocol"],
+			"timeout":                 funcInfo["timeout"],
+		},
+	}
+	return d.Set("apig", result)
+}
+
+func setDedicatedApigEventData(d *schema.ResourceData, eventData map[string]interface{}) error {
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: the repetition with `setSharedApigEventData` is too high, and it is recommended to combine them

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 5]
```diff
@@ -619,6 +837,10 @@ func resourceFunctionGraphTriggerDelete(d *schema.ResourceData, meta interface{}
 
 	urn := d.Get("function_urn").(string)
 	triggerType := d.Get("type").(string)
+
+	if triggerType == apigTrigger || triggerType == dedicatedApigTrigger {
+
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: empty ?

**Comment by chengxiangdong** [comment]
is the coding job finished?

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:75]: yeah, I forgot

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: These codes has been removed.

**Comment by Lance52259** [comment]
Just finished now, I forgot **WIP** marker.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 6]
```diff
@@ -265,6 +278,59 @@ func kafkaSchemaResource() *schema.Resource {
...
+					"IAM", "APP", "NONE",
+				}, false),
+				Default: "IAM",
+			},
+			"request_protocol": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"HTTP", "HTTPS",
```
Comment (ShiChangkuo) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:319]: The valid values does not include **BOTH**, please check it

Comment (Lance52259) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:319]: fixed in **6ff2c07**.

**Pull request merged**